README.md
# Plug.AccessLog

Plug for writing access logs.

## Setup

To use the plug in your projects, edit your mix.exs file and add the project as a dependency:

```elixir
defp deps do
  [
    # ...
    {:plug_accesslog, "~> 0.15.0"},
    # ...
  ]
end
```

## Usage

Add the plug to your plug pipeline/router:

```elixir
defmodule AppRouter do
  use Plug.Router

  plug Plug.AccessLog,
    format: :clf,
    file: "/path/to/your/logs/access.log"

  plug :match
  plug :dispatch

  get "/hello" do
    send_resp(conn, 200, "world")
  end

  match _ do
    send_resp(conn, 404, "oops")
  end
end
```

__Note__: The usage examples apply to a usecase where your are using `plug` directly without any framework. Using the `plug Plug.AccessLog` line in a framework based on `plug` should be no problem. Please refer to your frameworks individual documentation or source to find a suitable place.

### WAL Configuration

All log messages that will be written to a file are collected in a WAL process
before actual writing. The messages will be fetched in a configurable interval
to be written to the logfiles:

```elixir
config :plug_accesslog,
  :wal,
    flush_interval: 100
```

The time is configured as "milliseconds between writing and flushing". The default value is `100` milliseconds.

### Custom Formatters

If you want to extend the formatting capabilities or replace existing ones you can define a custom formatter pipeline to use:

```elixir
defmodule CustomFormatter do
  @behaviour Plug.AccessLog.Formatter

  def format(format, conn) do
    # manipulate to your liking
    format
  end
end

defmodule Router do
  use Plug.Router

  plug Plug.AccessLog,
    format: :clf,
    formatters: [CustomFormatter, Plug.AccessLog.DefaultFormatter],
    file: "/path/to/your/logs/access.log"
end
```

If you do not configure a list of formatters only the `DefaultFormatter` will be used. If you define an empty list then no formatting will take place.

All formatters are called in the order they are defined in.

### File Configuration

There are two ways to define the file you want log entries to be written to:

```elixir
defmodule Router do
  use Plug.Router

  plug Plug.AccessLog, file: "/static/configuration.log"
  plug Plug.AccessLog, file: {:system, "SYS_ENV_VAR_WITH_FILE_PATH"}
  plug Plug.AccessLog, file: {:system, "SYS_ENV_VAR", "/path/to/default.log"}
end
```

### Do Not Log Filter

To filter the requests before logging you can configure a "do not log" filter function:

```elixir
defmodule LogFilter do
  def dontlog?(%{request_path: "/favicon.ico"}), do: true
  def dontlog?(_conn), do: false
end

defmodule Router do
  use Plug.Router

  plug Plug.AccessLog,
    dontlog: &LogFilter.dontlog?/1,
    format: :clf,
    file: "/path/to/your/logs/access.log"
end
```

If the function you pass to the plug returns `true` the request will not be logged.

### Logging Functions

To have the parsed log message sent to a logging function instead of writing it to a file you can configure a logging function:

```elixir
defmodule InfoLogger do
  def log(msg), do: Logger.log(:info, msg)
end

defmodule Router do
  use Plug.Router

  plug Plug.AccessLog,
    format: :clf,
    fun: &InfoLogger.log/1
end
```

If a logging function is configured the configured file (if any) will be ignored.

## Log Format

The default format is [CLF](http://en.wikipedia.org/wiki/Common_Log_Format).

### Available formats

Besides a self defined format you can use one of the predefined aliases:

```
:agent
> %{User-Agent}i
> curl/7.35.0

:clf
> %h %l %u %t "%r" %>s %b
> 127.0.0.1 - - [10/Jan/2015:14:46:18 +0100] "GET / HTTP/1.1" 200 31337

:clf_vhost
> %v %h %l %u %t "%r" %>s %b
> www.example.com 127.0.0.1 - - [10/Jan/2015:14:46:18 +0100] "GET / HTTP/1.1" 200 31337

:combined
> %h %l %u %t "%r" %>s %b "%{Referer}i" "%{User-Agent}i"
> 127.0.0.1 - - [22/Jan/2015:19:33:58 +0100] "GET / HTTP/1.1" 200 2 "http://www.example.com/previous_page" "curl/7.35.0"

:combined_vhost
> %v %h %l %u %t "%r" %>s %b "%{Referer}i" "%{User-Agent}i"
> www.example.com 127.0.0.1 - - [22/Jan/2015:19:33:58 +0100] "GET / HTTP/1.1" 200 2 "http://www.example.com/previous_page" "curl/7.35.0"

:referer
> %{Referer}i -> %U
> http://www.example.com/previous_page -> /
```

For more details about each formatting token and potential modifications please refer to the `Plug.AccessLog.DefaultFormatter` module.

## Benchmarking

A small utility script is provided to check how long it might take to process requests and write the log messages to your disk:

```shell
mix run utils/bench.exs
```

This call will send of a total of 10k requests and wait for them to be written to the disk.

Looking at the data written to `utils/bench.log` might give a hint at what overhead the log writing is introducing. As with all "benchmarks" of any kind: take the measurements with a pinch of salt and run them in dozens of different conditions yourself.

## License

[Apache License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0)


CHANGELOG.md
# Changelog

## v0.16.0-dev

- Backwards incompatible changes
    - Minimum required elixir version is now `~> 1.9`

## v0.15.0 (2020-07-12)

- Enhancements
    - Protocol for `%r` is now logged as reported by `Plug.Conn.get_http_protocol/1`
    - System environment configuration can set an optional default value to be used if the environment variable is unset

- Backwards incompatible changes
    - Minimum required elixir version is now `~> 1.7`

## v0.14.0 (2016-11-25)

- Additional formatting directives
    - `%{VARNAME}e` - Environment variable contents

- Backwards incompatible changes
    - All modules of the `DefaultFormatter` have been changed to not append to a passed message anymore but return `iodata`. The public method has been renamed from `append` to `format` to reflect this change.

## v0.13.0 (2016-07-13)

- Backwards incompatible changes
    - Minimum required elixir version is now `~> 1.3`
    - Minimum required version of :timex raised to `~> 3.0`

## v0.12.0 (2016-06-25)

- Backwards incompatible changes
    - Minimum required elixir version is now `~> 1.2`
    - Minimum required erlang version is now `~> 18.0`

## v0.11.0 (2016-03-25)

- Enhancements
    - Logfiles can be configured using system environment variables
    - Logs are written in batches to the output file (configurable, default `100ms`)

- Backwards incompatible changes
    - `:local_time` is now internally stored as a `Timex.DateTime` struct
    - Minimum required version of :timex raised to `~> 2.0` ([#12](https://github.com/mneudert/plug_accesslog/pull/12))

## v0.10.0 (2016-01-06)

- Additional formatting directives
    - `%P` - The process ID that serviced the request
    - `%{UNIT}T` - Time taken to serve the request in the given UNIT

- Backwards incompatible changes
    - Minimum required version of :timex raised to `~> 1.0`

## v0.9.1 (2015-09-29)

- Enhancements
    - Application :tzdata (for recent :timex/:tzdata versions) is started automatically

- Backwards incompatible changes
    - Minimum required version of :timex raised to `~> 0.19`
    - Minimum required version of :tzdata raised to `=> 0.5.1`

## v0.9.0 (2015-08-15)

- Backwards incompatible changes
    - Minimum required version of :plug raised to `~> 1.0`

## v0.8.0 (2015-08-08)

- Backwards incompatible changes
    - Minimum required version of :plug raised to `~> 0.14`

## v0.7.0 (2015-07-19)

- Enhancements
    - Allows passing custom formatters
    - Errors when open logfiles are logged as errors using Logger
    - Log writing is now handled by a GenEvent handler to avoid crashing

- Additional formatting directives
    - `%a` - Remote IP-address
    - `%D` - Time taken to serve the request (microseconds)
    - `%M` - Time taken to serve the request (milliseconds)
    - `%q` - Query string (prepended with "?" or empty string)
    - `%T` - Time taken to serve the request (full seconds)
    - `%V` - Server name (canonical)

- Backwards incompatible changes
    - `Plug.AccessLog.Formatter` has been renamed to `Plug.AccessLog.DefaultFormatter`

## v0.6.0 (2015-05-23)

- Enhancements
    - Allows passing a dontlog-function to skip requests from logging
    - Dependencies not used in production builds are marked as optional

- Additional formatting directives
    - `%{VARNAME}C` - Cookie sent by the client
    - `%m` - Request method
    - `%{VARNAME}o` - Header line sent by the server

## v0.5.1 (2015-05-11)

- Bug fixes
    - Properly accesses headers using lowercase (@c-rack)

## v0.5.0 (2015-02-24)

- Enhancements
    - Allows logging to functions instead of files
    - Properly ignores unopenable logfiles (e.g. "..")
    - Recreates logfile if necessary (e.g. after moved by logrotate)

- Additional formatting directives
    - `%%` - Percentage sign
    - `%B` - Size of response in bytes. Outputs "0" when no bytes are sent.

## v0.4.1 (2015-02-03)

- Bug fixes
    - Properly handles charlist responses (@chvanikoff)
    - Reopens logfile IO device if pid is not alive anymore (@chvanikoff)

## v0.4.0 (2015-01-31)

- Enhancements
    - Provides "Agent Log Format" alias
    - Provides "Combined Log Format" alias
    - Provides "Combined Log Format with VHost" alias
    - Provides "Referer Log Format" alias

## v0.3.0 (2015-01-25)

- Enhancements
    - Logs username from basic authentication
    - Provides access to vhost (domain) for logging
    - Raises plug version requirement to `~> 0.10`

## v0.2.0 (2015-01-18)

- Enhancements
    - Supports full [CLF](http://en.wikipedia.org/wiki/Common_Log_Format)

## v0.1.0 (2015-01-15)

- Initial Release


mix.exs
defmodule Plug.AccessLog.MixProject do
  use Mix.Project

  @url_github "https://github.com/mneudert/plug_accesslog"

  def project do
    [
      app: :plug_accesslog,
      name: "Plug.AccessLog",
      version: "0.16.0-dev",
      elixir: "~> 1.9",
      deps: deps(),
      description: "Plug for writing access logs",
      dialyzer: dialyzer(),
      docs: docs(),
      package: package(),
      preferred_cli_env: [
        coveralls: :test,
        "coveralls.detail": :test
      ],
      test_coverage: [tool: ExCoveralls]
    ]
  end

  def application do
    [
      extra_applications: [:logger, :plug],
      mod: {Plug.AccessLog.Application, []}
    ]
  end

  defp dialyzer do
    [
      flags: [
        :error_handling,
        :race_conditions,
        :underspecs,
        :unmatched_returns
      ],
      plt_core_path: "plts",
      plt_file: {:no_warn, "plts/dialyzer.plt"}
    ]
  end

  defp deps do
    [
      {:credo, "~> 1.0", only: :dev, runtime: false},
      {:dialyxir, "~> 1.0", only: :dev, runtime: false},
      {:ex_doc, ">= 0.0.0", only: :dev, runtime: false},
      {:excoveralls, "~> 0.14.0", only: :test, runtime: false},
      {:plug, "~> 1.6"},
      {:timex, "~> 3.0"}
    ]
  end

  defp docs do
    [
      extras: ["CHANGELOG.md", "README.md"],
      main: "readme",
      source_ref: "master",
      source_url: @url_github
    ]
  end

  defp package do
    %{
      files: ["CHANGELOG.md", "LICENSE", "mix.exs", "README.md", "lib"],
      licenses: ["Apache 2.0"],
      links: %{"GitHub" => @url_github}
    }
  end
end
